File: src/rules/RulesXmlHandler.cpp
#include <QtCore/QObject>

#include "RulesXmlHandler.h"
#include "RootNode.h"
#include "ActionNode.h"
#include "StorageNode.h"
#include "EmotionNode.h"
#include "ConditionNode.h"
#include "ConditionPerceptionNode.h"
#include "ConditionEmotionNode.h"
#include "ConditionDataMiningNode.h"
#include "../Brain.h"

namespace CnotiMind
{

	RulesXmlHandler::RulesXmlHandler( Brain* brain) :
		QXmlDefaultHandler(),
		_rootNode( NULL ),
		_currentNode( NULL ),
		_parentNode( NULL ),
		_brain( brain ),
		_parentObject( new QObject(0) )
	{
	}

	RulesXmlHandler::~RulesXmlHandler()
	{
		delete _parentObject;
	}


	bool RulesXmlHandler::startElement ( const QString & namespaceURI, const QString & localName, const QString & qName, const QXmlAttributes & atts )
	{
		Q_UNUSED( namespaceURI );
		Q_UNUSED( localName );

		if( QString::compare( qName, "Rules", Qt::CaseInsensitive ) == 0 )
		{
			return createRootNode( atts );
		}
		else if( QString::compare( qName, "Condition", Qt::CaseInsensitive ) == 0 )
		{
			return createConditionNode( atts );
		}
		else if( QString::compare( qName, "Action", Qt::CaseInsensitive ) == 0 )
		{
			return createActionNode( atts );
		}
		else if( QString::compare( qName, "Storage", Qt::CaseInsensitive ) == 0 )
		{
			return createStorageNode( atts );
		}
		else if( QString::compare( qName, "Emotion", Qt::CaseInsensitive ) == 0 )
		{
			return createEmotionNode( atts );
		}

		return false;
	}

	/*
		When an element is closed the currentNode and parentNode are updated
	*/
	bool RulesXmlHandler::endElement( const QString & namespaceURI, const QString & localName, const QString & qName )
	{
		Q_UNUSED( namespaceURI );
		Q_UNUSED( localName );
		Q_UNUSED( qName );

		if( _rootNode == _currentNode )
		{
			_currentNode = NULL;
			return true;
		}

		_currentNode = _parentNode;
		_parentNode = qobject_cast<RuleNode*>( _currentNode->parent() );

		return true;
	}

	/*
		Return the RootNode from the rules. Who calls this will be responsible for freeing
		the memory of the Rules.
	*/
	RuleNode* RulesXmlHandler::rootNode() const
	{
		// removes the handler has parent, so who calls now must be responsible for deleting all nodes
		_rootNode->setParent( 0 );

		return _rootNode;
	}

	/*
		Creates a root node. It can only exist 1 root node.
	*/
	bool RulesXmlHandler::createRootNode(  const QXmlAttributes & atts )
	{
		Q_UNUSED( atts );

		// it can only exist 1 root node.
		if( _rootNode != NULL )
		{
			return false;
		}

		_rootNode = new RootNode( _brain, _parentObject );
		_currentNode = _rootNode;

		return true;
	}

	bool RulesXmlHandler::createActionNode(  const QXmlAttributes & atts )
	{
		if( _rootNode == NULL || _currentNode == NULL )
		{
			return false;
		}

		QString key = atts.value( "key" );
		QString value = atts.value( "value" );

		_parentNode = _currentNode;
		_currentNode =  new ActionNode( key, value, _brain, _parentNode );

		return true;

	}

	bool RulesXmlHandler::createStorageNode(  const QXmlAttributes & atts )
	{
		if( _rootNode == NULL || _currentNode == NULL )
		{
			return false;
		}

		QString key = atts.value( "key" );
		QString value = atts.value( "value" );
		Brain::MemoryType memory = Brain::translateMemoryType( atts.value( "memory" ) );

		_parentNode = _currentNode;
		_currentNode =  new StorageNode( key, value, memory, _brain, _parentNode );

		return true;
	}

	bool RulesXmlHandler::createEmotionNode(  const QXmlAttributes & atts )
	{
		if( _rootNode == NULL || _currentNode == NULL )
		{
			return false;
		}

		QString key = atts.value( "key" );
		QString value = atts.value( "value" );

		bool okMin;
		int min = atts.value( "min" ).toInt(&okMin);
		bool okMax;
		int max = atts.value( "max" ).toInt(&okMax);

		if(!okMax)
		{
			max = INT_MAX;
		}
		if(!okMin)
		{
			min = INT_MIN;
		}

		_parentNode = _currentNode;
		_currentNode =  new EmotionNode( key, value, max, min, _brain, _parentNode );

		return true;
	}

	/*

	*/
	bool RulesXmlHandler::createConditionNode(  const QXmlAttributes & atts )
	{
		if( _rootNode == NULL || _currentNode == NULL )
		{
			return false;
		}

		if( atts.index( "perception") != -1)
		{
			return createConditionPerceptionNode( atts );
		}
		else if( atts.index( "datamining" ) != -1)
		{
			return createConditionDataMiningNode( atts );
		}
		else if( atts.index( "emotion" ) != -1 )
		{
			return createConditionEmotionNode( atts );
		}

		return false;
	}

	/*

	*/
	bool RulesXmlHandler::createConditionPerceptionNode(  const QXmlAttributes & atts )
	{
		QString key = atts.value( "perception" );
		QString value = atts.value( "value" );
		QString opString = atts.value( "operator" );
		ConditionNode::ConditionOperator op = ConditionNode::translateConditionOperator( opString );

		_parentNode = _currentNode;
		_currentNode =  new ConditionPerceptionNode( key, value, op, _brain, _parentNode );

		return true;

	}

	/*

	*/
	bool RulesXmlHandler::createConditionEmotionNode(  const QXmlAttributes & atts )
	{
		QString key = atts.value( "emotion" );
		QString value = atts.value( "value" );
		ConditionNode::ConditionOperator op = ConditionNode::translateConditionOperator( atts.value( "operator" ) );

		_parentNode = _currentNode;
		_currentNode =  new ConditionEmotionNode( key, value, op, _brain, _parentNode );

		return true;
	}

	/*

	*/
	bool RulesXmlHandler::createConditionDataMiningNode(  const QXmlAttributes & atts )
	{
		QString key = atts.value( "event" );
		QString value = atts.value( "value" );
		ConditionNode::ConditionOperator op = ConditionNode::translateConditionOperator( atts.value( "operator" ) );
		ConditionNode::DataMiningOperation opDataMining = ConditionNode::translateDataMiningOperator( atts.value( "datamining" ) );
		Brain::MemoryType memory = Brain::translateMemoryType( atts.value( "memory" ) );

		_parentNode = _currentNode;
		_currentNode =  new ConditionDataMiningNode( key, value, op, opDataMining, memory, _brain, _parentNode );

		return true;
	}


}

File: src_ios/CnotiMind/RulesXmlHandler.m
//
//  RulesXmlHandler.m
//  SingleEmotion
//
//  Created by Gonçalo Rodrigues on 8/17/11.
//  Copyright 2011 Universidade de Coimbra. All rights reserved.
//

#import "RulesXmlHandler.h"
#import "RootNode.h"
#import "ActionNode.h"
#import "StorageNode.h"
#import "EmotionNode.h"
#import "ConditionNode.h"
#import "ConditionPerceptionNode.h"
#import "ConditionEmotionNode.h"
#import "ConditionDataMiningNode.h"
#import "ConditionVariableNode.h"
#import "DataMiningNode.h"

#import "CnotiMind.h"
#import "Brain.h"

@implementation RulesXmlHandler

@synthesize rootNode = _rootNode;
@synthesize currentNode = _currentNode;

- (id)init
{
    self = [super init];
    if (self) {

        // Initialization code here.
        _rootNode = NULL;
        _currentNode = NULL;
        _parentNode = NULL;
//        _brain = [[Brain alloc] init];
        _parentObject = [[RuleNode alloc] init];
        _line = 0;
    }
    return self;
}

- (id)initWithBrain:(Brain*)aBrain
{
    self = [super init];
    if (self) {
        // Initialization code here.
        _rootNode = NULL;
        _currentNode = NULL;
        _parentNode = NULL;
        _brain = aBrain;
        _parentObject = [[RuleNode alloc] init];
        _line = 0;

    }
    return self;
}


- (BOOL) startElement:(NSString*)aNamespaceURI localName:(NSString*)aLocalName qName:(NSString*)aQName atts:(GDataXMLElement*)atts
{
    _line++;
    
    if( [aQName isEqualToString:@"Rules"] )
    {
        return [self createRootNode:atts];
    }
    if( [aQName isEqualToString:@"Condition"] )
    {
        return [self createConditionNode:atts];
    }
    if( [aQName isEqualToString:@"Action"] )
    {
        return [self createActionNode:atts];
    }
    if( [aQName isEqualToString:@"Storage"] )
    {
        return [self createStorageNode:atts];
    }
    if( [aQName isEqualToString:@"Emotion"] )
    {
        return [self createEmotionNode:atts];
    }
    if( [aQName isEqualToString:@"DataMining"] )
    {
        return [self createDataMiningNode:atts];
    }
    if( [aQName isEqualToString:@"MathOperation"] )
    {
        return [self createMathOperationNode:atts];
    }
    if( [aQName isEqualToString:@"Property"] )
    {
        return [self createPropertyNode:atts];
    }
    if( [aQName isEqualToString:@"Delete"] )
    {
        return [self createDeleteNode:atts];
    }
    if( [aQName isEqualToString:@"ClearMemory"] )
    {
        return [self createClearMemoryNode:atts];
    }
    if( [aQName isEqualToString:@"Random"] )
    {
        return [self createRandomNode:atts];
    }
    
    DLog(@"[RulesXmlHandler::startElement] Invalid element: %@",aQName);
    
    return false;
}

/*
 When an element is closed the currentNode and parentNode are updated
*/
- (BOOL) endElement:(NSString*)aNamespaceURI localName:(NSString*)aLocalName qName:(NSString*)aQName
{
    if( _rootNode == _currentNode )
    {
        _currentNode = NULL;
        return true;
    }
    
    _currentNode = _parentNode;
    _parentNode = (RuleNode*)_currentNode.parent;
    
    return true;
}


- (BOOL) createRootNode:(GDataXMLElement*)atts
{
    
    // it can only exist 1 root node.
    if( _rootNode != NULL )
    {
        return false;
    }
    
    _rootNode = [[RootNode alloc] initWithBrainAndParent: _brain parent:_parentObject];
    _currentNode = _rootNode;
    
    return true;
}

- (BOOL) createActionNode:(GDataXMLElement*)atts
{
    if( _rootNode == NULL || _currentNode == NULL )
    {
        return false;
    }

    NSString* name = [[atts attributeForName:@"name"] stringValue];
    NSString* value = [[atts attributeForName:@"value"] stringValue];
    NSString* probability = [[atts attributeForName:@"probability"] stringValue];
    
    if( [probability length]==0 )
    {
        probability = @"1";
    }
    
    _parentNode = _currentNode;
    _currentNode = [[ActionNode alloc] initWithNameAndValueAndBrainAndParent: name value:value brain:_brain parent:_parentNode];
    
    return true;
}

- (BOOL) createStorageNode:(GDataXMLElement*)atts
{
    
    if( _rootNode == NULL || _currentNode == NULL )
    {
        return false;
    }
    
    NSString* event = [[atts attributeForName:@"event"] stringValue];
    NSString* value = [[atts attributeForName:@"value"] stringValue];
    NSString* clear = [[atts attributeForName:@"clear"] stringValue];
    
    NSString* memory = [[atts attributeForName:@"memory"] stringValue];

    enum MemoryType memoryType;
    memoryType = [CnotiMind translateMemoryType: memory];
    
    _parentNode = _currentNode;
    
    if ([clear isEqualToString:@"yes"]) {
        _currentNode = [[StorageNode alloc] initWithClearAndMemoryAndBrainAndParent:true memory:memoryType brain:_brain parent:_parentNode];
    }
    else {
        _currentNode = [[StorageNode alloc] initWithEventAndValueAndMemoryAndBrainAndParent:event value:value memory:memoryType brain:_brain parent:_parentNode];
    }
    
    return true;
    
}

- (BOOL) createEmotionNode:(GDataXMLElement*)atts
{
    
    
    
    if( _rootNode == NULL || _currentNode == NULL )
    {
        return false;
    }
    
    NSString* emotion = [[atts attributeForName:@"name"] stringValue];
    NSString* value = [[atts attributeForName:@"increase"] stringValue];
    BOOL okMin = [CnotiMind isNumeric: [[atts attributeForName:@"min"] stringValue]];
    double min = [[[atts attributeForName:@"min"] stringValue] floatValue];
    BOOL okMax = [CnotiMind isNumeric: [[atts attributeForName:@"max"] stringValue]];
    double max = [[[atts attributeForName:@"max"] stringValue] floatValue];
        
    if (!okMin) {
        min = INT_MIN;
    }
    if (!okMax) {
        max = INT_MAX;
    }
    _parentNode = _currentNode;
    _currentNode = [[EmotionNode alloc] initWithEmotionAndValueAndMaxAndMinAndBrainAndParent:emotion value:value max:max min:min brain:_brain parent:_parentNode];
    
    return true;
}

- (BOOL) createDataMiningNode:(GDataXMLElement*)atts
{
    if( _rootNode == NULL || _currentNode == NULL )
    {
        return false;
    }
    
    NSString* event = [[atts attributeForName:@"event"] stringValue];
    NSString* value = [[atts attributeForName:@"value"] stringValue];
    NSString* op = [[atts attributeForName:@"operator"] stringValue];
    NSString* opDataMining = [[atts attributeForName:@"operation"] stringValue];
    NSString* variable = [[atts attributeForName:@"variable"] stringValue];
//TODO
    //    NSString* position = [[atts attributeForName:@"position"] stringValue];
    NSString* memory = [[atts attributeForName:@"memory"] stringValue];

    enum ConditionOperator opType;
    opType = [CnotiMind translateConditionOperator: op];
    
    enum DataMiningOperation opDataMiningType;
    opDataMiningType = [CnotiMind translateDataMiningOperator: opDataMining];
    
    enum MemoryType memoryType;
    memoryType = [CnotiMind translateMemoryType: memory];

    
    _parentNode = _currentNode;
    _currentNode = [[DataMiningNode alloc] initWithEventAndValueAndOperatorAndBrainAndParent:event value:value operator:opDataMiningType memory:memoryType variable:variable brain:_brain parent:_parentNode];
    
    return true;
}

- (BOOL) createConditionNode:(GDataXMLElement*)atts
{
    if( _rootNode == NULL || _currentNode == NULL )
    {
        return false;
    }
    
    NSString* typeCondition = [[atts attributeForName:@"type"] stringValue];

    
    if ([typeCondition isEqualToString:@"perception"]) {
        return [self createConditionPerceptionNode:atts];
    }
    else if ([typeCondition isEqualToString:@"datamining"]) {
        return [self createConditionDataMiningNode:atts];
    }
    else if ([typeCondition isEqualToString:@"emotion"]) {
        return [self createConditionEmotionNode:atts];
    }
    else if ([typeCondition isEqualToString:@"variable"]) {
        return [self createConditionVariableNode:atts];
    }
    else if ([typeCondition isEqualToString:@"property"]) {
        return [self createConditionPropertyNode:atts];
    }
    
    DLog(@"[RulesXmlHandler::createConditionNode] Invalid condition node type: %@", typeCondition);
    
    return false;    
}


- (BOOL) createConditionPerceptionNode:(GDataXMLElement*)atts
{
    NSString* perception = [[atts attributeForName:@"perception"] stringValue];
    NSString* value = [[atts attributeForName:@"value"] stringValue];
    NSString* opCondition = [[atts attributeForName:@"operator"] stringValue];

    enum ConditionOperator opConditionType;
    opConditionType = [CnotiMind translateConditionOperator: opCondition];
    
    _parentNode = _currentNode;
    _currentNode = [[ConditionPerceptionNode alloc] initWithPerceptionAndValueAndOperatorAndBrainAndParent:perception value:value operator:opConditionType brain:_brain parent:_parentNode];
    
    return true;
}


- (BOOL) createConditionVariableNode:(GDataXMLElement*)atts
{
    NSString* variable = [[atts attributeForName:@"variable"] stringValue];
    NSString* value = [[atts attributeForName:@"value"] stringValue];
    NSString* opCondition = [[atts attributeForName:@"operator"] stringValue];
    
    enum ConditionOperator opConditionType;
    opConditionType = [CnotiMind translateConditionOperator: opCondition];
    
    _parentNode = _currentNode;
    _currentNode = [[ConditionVariableNode alloc] initWithVariableAndValueAndOperatorAndBrainAndParent:variable value:value operator:opConditionType brain:_brain parent:_parentNode];
    
    return true;    
}

- (BOOL) createConditionPropertyNode:(GDataXMLElement*)atts
{
//    TODO
    return false;
}

- (BOOL) createConditionEmotionNode:(GDataXMLElement*)atts
{
    NSString* emotion = [[atts attributeForName:@"emotion"] stringValue];
    NSString* value = [[atts attributeForName:@"value"] stringValue];
    NSString* opCondition = [[atts attributeForName:@"operator"] stringValue];
    
    enum ConditionOperator opConditionType;
    opConditionType = [CnotiMind translateConditionOperator: opCondition];
    
    _parentNode = _currentNode;
    _currentNode = [[ConditionVariableNode alloc] initWithEmotionAndValueAndOperatorAndBrainAndParent:emotion value:value operator:opConditionType brain:_brain parent:_parentNode];
    
    return true;  
}

- (BOOL) createConditionDataMiningNode:(GDataXMLElement*)atts
{
    NSString* event = [[atts attributeForName:@"event"] stringValue];
    NSString* value = [[atts attributeForName:@"value"] stringValue];
    NSString* operator = [[atts attributeForName:@"operator"] stringValue];
    NSString* operation = [[atts attributeForName:@"operation"] stringValue];
    NSString* variable = [[atts attributeForName:@"variable"] stringValue];
    NSString* compareValue = [[atts attributeForName:@"compareValue"] stringValue];
    NSString* memory = [[atts attributeForName:@"memory"] stringValue];

    enum DataMiningOperation operationType;
    operationType = [CnotiMind translateDataMiningOperator: operation];
    
    enum ConditionOperator operatorType;
    operatorType = [CnotiMind translateConditionOperator: operator];
    
    enum MemoryType memoryType;
    memoryType = [CnotiMind translateMemoryType: memory];
    
    _parentNode = _currentNode;
    _currentNode = [[ConditionDataMiningNode alloc] initWithKeyAndValueAndOperatorAndOperationAndMemoryAndVariableAndCompareValueBrainAndParent:event 
                                                                                                                                          value:value
                                                                                                                                       operator:operatorType
                                                                                                                                      operation:operationType 
                                                                                                                                         memory:memoryType 
                                                                                                                                       variable:variable 
                                                                                                                                   compareValue:compareValue 
                                                                                                                                          brain:_brain 
                                                                                                                                         parent:_parentNode];
    
    return true;     
}

- (BOOL) createMathOperationNode:(GDataXMLElement*)atts
{
    if( _rootNode == NULL || _currentNode == NULL )
    {
        return false;
    }
    
//    NSString* operation = [[atts attributeForName:@"name"] stringValue];
//    NSString* value = [[atts attributeForName:@"value"] stringValue];
//    NSString* variable = [[atts attributeForName:@"variable"] stringValue];
//    NSString* result = [[atts attributeForName:@"result"] stringValue];
//    
//    enum MathOperation mathOperationType;
//    mathOperationType = [CnotiMind translateMathOperation:operation];
//    TODO
//    _currentNode = [MathOperationNode]
    
    return false;
}

- (BOOL) createPropertyNode:(GDataXMLElement*)atts
{
//    TODO    
    return false;   
}

- (BOOL) createDeleteNode:(GDataXMLElement*)atts
{
//    TODO    
    return false;       
}

- (BOOL) createClearMemoryNode:(GDataXMLElement*)atts
{
//    TODO    
    return false;   
}

- (BOOL) createRandomNode:(GDataXMLElement*)atts
{
//    TODO    
    return false;    
}



@end
